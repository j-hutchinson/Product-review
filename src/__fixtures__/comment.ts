import { Comment } from '../types';

export const singleComment: Comment = {
	name: 'Jack',
	email: 'something@aol.com',
	rating: 2,
	comment: 'adequate product'
};

export const multipleComments: Comment[] = [
	{
		name: 'Jack',
		email: 'somewhere@aol.com',
		rating: 3,
		comment: 'adequate product'
	},
	{
		name: 'Jonny',
		email: 'over@aol.com',
		rating: 5,
		comment: 'I love it!'
	},
	{
		name: 'Jamie',
		email: 'the@aol.com',
		rating: 1,
		comment: 'I hate it'
	},
	{
		name: 'Joanne',
		email: 'rainbow@aol.com',
		rating: 2,
		comment: "I don't know how to feel"
	}
];
