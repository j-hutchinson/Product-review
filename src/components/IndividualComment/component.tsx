import React from 'react';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { Comment } from '../../types';

interface Props {
    comment: Comment;
}

const StyledComment = styled.li`
    margin: 0 ${spacing.medium};
    padding: ${spacing.medium} 0;
`;

const StyledLabel = styled.div`
    padding: 2px 0;
`;

const StyledTopLabel = styled.div`
    display: flex;
    padding: 2px 0;
`;

const StyledEnd = styled.div`
    margin-left: auto;
`;

const IndividualComment = ({ comment: { name, rating, email, comment, timePosted } }: Props): JSX.Element => (
    <StyledComment>
        <StyledTopLabel>
            <div><strong>{`Name: `}</strong>{name}</div>
            <StyledEnd><strong>{`Date: `}</strong>{timePosted}</StyledEnd>
        </StyledTopLabel>
        <StyledLabel><strong>{`Rating: `}</strong>{rating}</StyledLabel>
        <StyledLabel><strong>{`Email: `}</strong>{email}</StyledLabel>
        <StyledLabel><strong>{`Comment: `}</strong>{comment}</StyledLabel>
    </StyledComment>
);

export default IndividualComment;