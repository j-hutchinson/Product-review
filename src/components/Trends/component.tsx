import React from 'react';
import { connect } from 'react-redux';
import { Doughnut } from 'react-chartjs-2';
import styled from 'styled-components'
import { ReduxState } from '../../types';

interface Props {
    ratings: number[]
}

const TrendContainer = styled.div`
    padding: 20px;
    border: 1px solid black;
    border-radius: 4px;
`;

export const StyledEmptyLabel = styled.div`
    text-align: center;
    margin: auto;
    padding-top: 40px;
`;

export const Trends = ({ ratings }) => {
    let one = 0; let two = 0; let three = 0; let four = 0; let five = 0;
    ratings.forEach(rating => {
        if (rating === 1) one++;
        if (rating === 2) two++;
        if (rating === 3) three++;
        if (rating === 4) four++;
        if (rating === 5) five++;
    })

    const data = {
        labels: ['1', '2', '3', '4', '5'],
        datasets: [
            {
                label: 'Ratings',
                backgroundColor: ['#FF6384', '#36A2EB', '#FFCE56', '#03FC28'],
                borderColor: 'black',
                borderWidth: 1,
                hoverBackgroundColor: 'rgba(255,99,132,0.4)',
                hoverBorderColor: 'rgba(255,99,132,1)',
                data: [one, two, three, four, five]
            }
        ]
    };

    return (
        <TrendContainer>
            {ratings.length ?
                <Doughnut
                    data={data}
                    width={100}
                    height={50}
                    options={{ maintainAspectRatio: false }}
                /> :
                <StyledEmptyLabel>No data available, try leaving a comment!</StyledEmptyLabel>
            }
        </TrendContainer>
    )
}

export const mapStateToProps = (state: ReduxState): Props => ({
    ratings: state.allComments.map(comment => comment.rating)
})

export default connect(mapStateToProps, null)(Trends);