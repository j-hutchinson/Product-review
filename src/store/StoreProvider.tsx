import React, { createContext, useReducer } from 'react';
import { ActionType, StoreType } from '../types';
import reducer, { initialState } from './reducers';

interface Props {
    children: JSX.Element | JSX.Element[];
}

interface ContextType {
    state: StoreType;
    dispatch: (action: ActionType) => void;
}

export const StoreContext = createContext<ContextType>(null);

export const StoreProvider = ({ children }: Props): JSX.Element => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return <StoreContext.Provider value={{ state, dispatch }}>{children}</StoreContext.Provider>;
};

