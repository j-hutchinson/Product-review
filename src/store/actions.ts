import { ActionType, Comment } from '../types';

const ADD_COMMENT = 'ADD_COMMENT';
const ADD_MULTIPLE_COMMENTS = 'ADD_MULTIPLE_COMMENTS';
const DELETE_COMMENTS = 'DELETE_COMMENTS';

const addComment = (comment: Comment): ActionType => {
	return { type: ADD_COMMENT, comment };
};

const addMultipleComments = (comments: Comment[]): ActionType => {
	return { type: ADD_MULTIPLE_COMMENTS, comments };
};

const deleteComments = (): ActionType => {
	return { type: DELETE_COMMENTS };
};

export {
	ADD_COMMENT,
	ADD_MULTIPLE_COMMENTS,
	DELETE_COMMENTS,
	addComment,
	addMultipleComments,
	deleteComments
};
