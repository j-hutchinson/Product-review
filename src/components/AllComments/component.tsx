import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { Comment, ReduxState } from '../../types';
import IndividualComment from '../IndividualComment/component';

interface Props {
    allComments: Comment[]
}

const StyledCommentsContainer = styled.div`
    border: 1px solid black;
    border-radius: 4px;
    padding: ${spacing.medium};
`;

const StyledStrong = styled.strong`
    padding-left: ${spacing.medium};
    font-size: 18px;
`;

export const StyledEmptyLabel = styled.p`
    padding-left: ${spacing.medium};
`;

export const AllComments = ({ allComments }: Props): JSX.Element => (
    <StyledCommentsContainer>
        <StyledStrong>Comments:</StyledStrong>
        {allComments.length ?
            allComments.map((comment: Comment, ind: number): JSX.Element =>
                <IndividualComment comment={comment} key={`${comment.name}${ind}`} />)
            :
            <StyledEmptyLabel>No comments available, be the first to leave a comment!</StyledEmptyLabel>
        }
    </StyledCommentsContainer>
);

export const mapStateToProps = (state: ReduxState): Props => ({
    allComments: state.allComments
})

export default connect(mapStateToProps, null)(AllComments);