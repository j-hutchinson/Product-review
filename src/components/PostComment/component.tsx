import React, { ChangeEvent, FormEvent, useState } from 'react'
import Rater from 'react-rater';
import 'react-rater/lib/react-rater.css';
import styled from 'styled-components'
import { addComment } from '../../store/actions/actions';
import store from '../../store/store';

export const StyledContainer = styled.form``;

const StyledItem = styled.div`
    display: grid;
    grid-template-rows: min-content 1fr;
    grid-gap: 4px;
    margin: 10px;
`;

export const StyledInput = styled.input`
   height: 24px;
   font-size: 14px;
`;

const StyledButtonWrapper = styled.div`
    display: flex;
    flex-direction: row-reverse;
    margin-right: 10px;
`;

const StyledSubmitInput = styled.input`
    background-color: #03fc28;
    border-radius: 4px;
    border: 1px solid black;
    font-size: 14px; 
    padding: 8px;
`;

const StyledRatingLabel = styled.p`
    padding-left: 20px;
    display: inline-block;
`;

const PostComment = (): JSX.Element => {
    const [name, setName] = useState<string>('');
    const [email, setEmail] = useState<string>('');
    const [rating, setRating] = useState<number>(0);
    const [comment, setComment] = useState<string>('');

    const onSubmission = (ev: FormEvent): void => {
        ev.preventDefault();
        store.dispatch(addComment({ name, email, rating, comment }));
        setName('');
        setEmail('');
        setRating(0);
        setComment('');
    }

    return (
        <StyledContainer onSubmit={onSubmission}>
            <StyledItem>
                <label htmlFor="name"> Name: </label>
                <StyledInput
                    id="name"
                    type="text"
                    value={name}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setName(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledItem>
                <label htmlFor="email"> Email: </label>
                <StyledInput
                    id="email"
                    type="email"
                    value={email}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setEmail(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledItem>Rating:
                <div>
                    <Rater
                        rating={rating}
                        total={5}
                        onRate={({ rating }): void => setRating(rating)}
                    />
                    <StyledRatingLabel>({rating}) stars</StyledRatingLabel>
                </div>
            </StyledItem>
            <StyledItem>
                <label htmlFor="comment"> Comment </label>
                <StyledInput
                    id="comment"
                    type="text"
                    value={comment}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setComment(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledButtonWrapper>
                <StyledSubmitInput type="submit" value="Submit" />
            </StyledButtonWrapper>
        </StyledContainer>
    );
}
export default PostComment