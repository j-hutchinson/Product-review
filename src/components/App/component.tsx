import React from 'react';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { mockData } from '../../data/mockData';
import { addMultipleComments, deleteComments } from '../../store/actions/actions';
import store from '../../store/store';
import AllComments from '../AllComments/component';
import PostComment from '../PostComment/component';
import Trends from '../Trends/component';

const StyledWrapper = styled.div`
  display: grid;
  grid-template-rows: max-content max-content;
  grid-gap: ${spacing.large};
  font-size: 14px;
  margin: 0;
  padding: ${spacing.medium};
  
  @media (min-width: 800px) {
    margin: ${spacing.medium} ${spacing.large};
    padding: ${spacing.xlarge};
  }
`;

const StyledSection = styled.div` 
  display: grid;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: 1fr;
  grid-gap: ${spacing.medium};
  
  @media (min-width: 800px) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 1fr;
    grid-gap: ${spacing.medium};
  }
`;

const Button = styled.button`
  width: 100px;
  margin: 0 ${spacing.small};
`;

const StyledButtonWrapper = styled.div`
  display: flex;
`;

/* 
* Helper buttons for testing purposes
* Adds/Removes mock comments to store to test functionality
*/

const App = (): JSX.Element => (
  <StyledWrapper>
    <StyledSection>
      <PostComment />
      <Trends />
    </StyledSection>
    <AllComments />
    <StyledButtonWrapper>
      <Button className="multiple-button" onClick={() => store.dispatch(addMultipleComments(mockData))}>Use mock data</Button>
      <Button className="clear-button" onClick={() => store.dispatch(deleteComments())}>Clear</Button>
    </StyledButtonWrapper>
  </StyledWrapper >
);


export default App;
