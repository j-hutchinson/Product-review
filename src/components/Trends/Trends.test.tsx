import { render } from '@testing-library/react';
import React from 'react';
import { multipleComments } from '../../__fixtures__/comment';
import { Trends } from './component';
import { StoreContext } from '../../store/StoreProvider'

describe('Trends component', () => {
    test('component matches snapshot with ratings', () => {
        expect.assertions(1);

        const wrapper = render(
            <StoreContext.Provider value={{ dispatch: jest.fn(), state: { allComments: multipleComments } }}>
                <Trends />
            </StoreContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('component matches snapshot with no ratings', () => {
        expect.assertions(1);

        const wrapper = render(
            <StoreContext.Provider value={{ dispatch: jest.fn(), state: { allComments: [] } }}>
                <Trends />
            </StoreContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });
});