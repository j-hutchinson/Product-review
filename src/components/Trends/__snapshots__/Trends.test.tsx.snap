// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trends component component matches snapshot with no ratings 1`] = `
<styled.div>
  <styled.div>
    No data available, try leaving a comment!
  </styled.div>
</styled.div>
`;

exports[`Trends component component matches snapshot with ratings 1`] = `
<styled.div>
  <Doughnut
    data={
      Object {
        "datasets": Array [
          Object {
            "backgroundColor": Array [
              "#FF6384",
              "#36A2EB",
              "#FFCE56",
              "#03FC28",
              "#E8E8E8",
            ],
            "borderColor": "#000000",
            "borderWidth": 1,
            "data": Array [
              1,
              1,
              1,
              2,
              1,
            ],
            "hoverBackgroundColor": Array [
              "#F7BCBC",
              "#C3D7F7",
              "#FFF8C7",
              "#D3FCC2",
              "#EDEDED",
            ],
            "hoverBorderColor": "#B5B5B5",
            "label": "Ratings",
          },
        ],
        "labels": Array [
          "1 star",
          "2 star",
          "3 star",
          "4 star",
          "5 star",
        ],
      }
    }
    height={50}
    options={
      Object {
        "maintainAspectRatio": false,
      }
    }
    width={100}
  />
</styled.div>
`;
